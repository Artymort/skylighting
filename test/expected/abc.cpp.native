[ [ ( PreprocessorTok , "#include " )
  , ( ImportTok , "<iostream>" )
  ]
, [ ( PreprocessorTok , "#include " )
  , ( ImportTok , "<vector>" )
  ]
, [ ( PreprocessorTok , "#include " )
  , ( ImportTok , "<string>" )
  ]
, [ ( PreprocessorTok , "#include " ) , ( ImportTok , "<set>" ) ]
, [ ( PreprocessorTok , "#include " )
  , ( ImportTok , "<cctype>" )
  ]
, []
, []
, [ ( KeywordTok , "typedef" )
  , ( NormalTok , " " )
  , ( BuiltInTok , "std::" )
  , ( NormalTok , "pair<" )
  , ( DataTypeTok , "char" )
  , ( NormalTok , "," )
  , ( DataTypeTok , "char" )
  , ( NormalTok , "> " )
  , ( DataTypeTok , "item_t" )
  , ( NormalTok , ";" )
  ]
, [ ( KeywordTok , "typedef" )
  , ( NormalTok , " " )
  , ( BuiltInTok , "std::" )
  , ( NormalTok , "vector<" )
  , ( DataTypeTok , "item_t" )
  , ( NormalTok , "> " )
  , ( DataTypeTok , "list_t" )
  , ( NormalTok , ";" )
  ]
, []
, [ ( DataTypeTok , "bool" )
  , ( NormalTok , " can_make_word(" )
  , ( AttributeTok , "const" )
  , ( NormalTok , " " )
  , ( BuiltInTok , "std::" )
  , ( NormalTok , "string& w, " )
  , ( AttributeTok , "const" )
  , ( NormalTok , " " )
  , ( DataTypeTok , "list_t" )
  , ( NormalTok , "& vals) {" )
  ]
, [ ( NormalTok , "    " )
  , ( BuiltInTok , "std::" )
  , ( NormalTok , "set<" )
  , ( DataTypeTok , "uint32_t" )
  , ( NormalTok , "> used;" )
  ]
, [ ( NormalTok , "    " )
  , ( ControlFlowTok , "while" )
  , ( NormalTok , " (used.size() < w.size()) {" )
  ]
, [ ( NormalTok , "        " )
  , ( AttributeTok , "const" )
  , ( NormalTok , " " )
  , ( DataTypeTok , "char" )
  , ( NormalTok , " c = toupper(w[used.size()]);" )
  ]
, [ ( NormalTok , "        " )
  , ( DataTypeTok , "uint32_t" )
  , ( NormalTok , " x = used.size();" )
  ]
, [ ( NormalTok , "        " )
  , ( ControlFlowTok , "for" )
  , ( NormalTok , " (" )
  , ( DataTypeTok , "uint32_t" )
  , ( NormalTok , " i = " )
  , ( DecValTok , "0" )
  , ( NormalTok , ", ii = vals.size(); i < ii; ++i) {" )
  ]
, [ ( NormalTok , "            " )
  , ( ControlFlowTok , "if" )
  , ( NormalTok , " (used.find(i) == used.end()) {" )
  ]
, [ ( NormalTok , "                " )
  , ( ControlFlowTok , "if" )
  , ( NormalTok
    , " (toupper(vals[i].first) == c || toupper(vals[i].second) == c) {"
    )
  ]
, [ ( NormalTok , "                    used.insert(i);" ) ]
, [ ( NormalTok , "                    " )
  , ( ControlFlowTok , "break" )
  , ( NormalTok , ";" )
  ]
, [ ( NormalTok , "                }" ) ]
, [ ( NormalTok , "            }" ) ]
, [ ( NormalTok , "        }" ) ]
, [ ( NormalTok , "        " )
  , ( ControlFlowTok , "if" )
  , ( NormalTok , " (x == used.size()) " )
  , ( ControlFlowTok , "break" )
  , ( NormalTok , ";" )
  ]
, [ ( NormalTok , "    }" ) ]
, [ ( NormalTok , "    " )
  , ( ControlFlowTok , "return" )
  , ( NormalTok , " used.size() == w.size();" )
  ]
, [ ( NormalTok , "}" ) ]
, []
, []
, [ ( DataTypeTok , "int" ) , ( NormalTok , " main() {" ) ]
, [ ( NormalTok , "    " )
  , ( DataTypeTok , "list_t" )
  , ( NormalTok , " vals{ {" )
  , ( StringTok , "'B'" )
  , ( NormalTok , "," )
  , ( StringTok , "'O'" )
  , ( NormalTok , "}, {" )
  , ( StringTok , "'X'" )
  , ( NormalTok , "," )
  , ( StringTok , "'K'" )
  , ( NormalTok , "}, {" )
  , ( StringTok , "'D'" )
  , ( NormalTok , "," )
  , ( StringTok , "'Q'" )
  , ( NormalTok , "}, {" )
  , ( StringTok , "'C'" )
  , ( NormalTok , "," )
  , ( StringTok , "'P'" )
  , ( NormalTok , "}, {" )
  , ( StringTok , "'N'" )
  , ( NormalTok , "," )
  , ( StringTok , "'A'" )
  , ( NormalTok , "}, {" )
  , ( StringTok , "'G'" )
  , ( NormalTok , "," )
  , ( StringTok , "'T'" )
  , ( NormalTok , "}, {" )
  , ( StringTok , "'R'" )
  , ( NormalTok , "," )
  , ( StringTok , "'E'" )
  , ( NormalTok , "}, {" )
  , ( StringTok , "'T'" )
  , ( NormalTok , "," )
  , ( StringTok , "'G'" )
  , ( NormalTok , "}, {" )
  , ( StringTok , "'Q'" )
  , ( NormalTok , "," )
  , ( StringTok , "'D'" )
  , ( NormalTok , "}, {" )
  , ( StringTok , "'F'" )
  , ( NormalTok , "," )
  , ( StringTok , "'S'" )
  , ( NormalTok , "}, {" )
  , ( StringTok , "'J'" )
  , ( NormalTok , "," )
  , ( StringTok , "'W'" )
  , ( NormalTok , "}, {" )
  , ( StringTok , "'H'" )
  , ( NormalTok , "," )
  , ( StringTok , "'U'" )
  , ( NormalTok , "}, {" )
  , ( StringTok , "'V'" )
  , ( NormalTok , "," )
  , ( StringTok , "'I'" )
  , ( NormalTok , "}, {" )
  , ( StringTok , "'A'" )
  , ( NormalTok , "," )
  , ( StringTok , "'N'" )
  , ( NormalTok , "}, {" )
  , ( StringTok , "'O'" )
  , ( NormalTok , "," )
  , ( StringTok , "'B'" )
  , ( NormalTok , "}, {" )
  , ( StringTok , "'E'" )
  , ( NormalTok , "," )
  , ( StringTok , "'R'" )
  , ( NormalTok , "}, {" )
  , ( StringTok , "'F'" )
  , ( NormalTok , "," )
  , ( StringTok , "'S'" )
  , ( NormalTok , "}, {" )
  , ( StringTok , "'L'" )
  , ( NormalTok , "," )
  , ( StringTok , "'Y'" )
  , ( NormalTok , "}, {" )
  , ( StringTok , "'P'" )
  , ( NormalTok , "," )
  , ( StringTok , "'C'" )
  , ( NormalTok , "}, {" )
  , ( StringTok , "'Z'" )
  , ( NormalTok , "," )
  , ( StringTok , "'M'" )
  , ( NormalTok , "} };" )
  ]
, [ ( NormalTok , "    " )
  , ( BuiltInTok , "std::" )
  , ( NormalTok , "vector<" )
  , ( BuiltInTok , "std::" )
  , ( NormalTok , "string> words{" )
  , ( StringTok , "\"A\"" )
  , ( NormalTok , "," )
  , ( StringTok , "\"BARK\"" )
  , ( NormalTok , "," )
  , ( StringTok , "\"BOOK\"" )
  , ( NormalTok , "," )
  , ( StringTok , "\"TREAT\"" )
  , ( NormalTok , "," )
  , ( StringTok , "\"COMMON\"" )
  , ( NormalTok , "," )
  , ( StringTok , "\"SQUAD\"" )
  , ( NormalTok , "," )
  , ( StringTok , "\"CONFUSE\"" )
  , ( NormalTok , "};" )
  ]
, [ ( NormalTok , "    " )
  , ( ControlFlowTok , "for" )
  , ( NormalTok , " (" )
  , ( AttributeTok , "const" )
  , ( NormalTok , " " )
  , ( BuiltInTok , "std::" )
  , ( NormalTok , "string& w : words) {" )
  ]
, [ ( NormalTok , "        " )
  , ( BuiltInTok , "std::" )
  , ( NormalTok , "cout << w << " )
  , ( StringTok , "\": \"" )
  , ( NormalTok , " << " )
  , ( BuiltInTok , "std::" )
  , ( NormalTok , "boolalpha << can_make_word(w,vals) << " )
  , ( StringTok , "\"." )
  , ( SpecialCharTok , "\\n" )
  , ( StringTok , "\"" )
  , ( NormalTok , ";" )
  ]
, [ ( NormalTok , "    }" ) ]
, []
, [ ( NormalTok , "}" ) ]
]
