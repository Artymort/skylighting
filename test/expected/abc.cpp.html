<div class="sourceCode"><pre class="sourceCode"><code class="sourceCode"><span class="pp" title="PreprocessorTok">#include </span><span class="im" title="ImportTok">&lt;iostream&gt;</span>
<span class="pp" title="PreprocessorTok">#include </span><span class="im" title="ImportTok">&lt;vector&gt;</span>
<span class="pp" title="PreprocessorTok">#include </span><span class="im" title="ImportTok">&lt;string&gt;</span>
<span class="pp" title="PreprocessorTok">#include </span><span class="im" title="ImportTok">&lt;set&gt;</span>
<span class="pp" title="PreprocessorTok">#include </span><span class="im" title="ImportTok">&lt;cctype&gt;</span>


<span class="kw" title="KeywordTok">typedef</span> <span class="bu" title="BuiltInTok">std::</span>pair&lt;<span class="dt" title="DataTypeTok">char</span>,<span class="dt" title="DataTypeTok">char</span>&gt; <span class="dt" title="DataTypeTok">item_t</span>;
<span class="kw" title="KeywordTok">typedef</span> <span class="bu" title="BuiltInTok">std::</span>vector&lt;<span class="dt" title="DataTypeTok">item_t</span>&gt; <span class="dt" title="DataTypeTok">list_t</span>;

<span class="dt" title="DataTypeTok">bool</span> can_make_word(<span class="at" title="AttributeTok">const</span> <span class="bu" title="BuiltInTok">std::</span>string&amp; w, <span class="at" title="AttributeTok">const</span> <span class="dt" title="DataTypeTok">list_t</span>&amp; vals) {
    <span class="bu" title="BuiltInTok">std::</span>set&lt;<span class="dt" title="DataTypeTok">uint32_t</span>&gt; used;
    <span class="cf" title="ControlFlowTok">while</span> (used.size() &lt; w.size()) {
        <span class="at" title="AttributeTok">const</span> <span class="dt" title="DataTypeTok">char</span> c = toupper(w[used.size()]);
        <span class="dt" title="DataTypeTok">uint32_t</span> x = used.size();
        <span class="cf" title="ControlFlowTok">for</span> (<span class="dt" title="DataTypeTok">uint32_t</span> i = <span class="dv" title="DecValTok">0</span>, ii = vals.size(); i &lt; ii; ++i) {
            <span class="cf" title="ControlFlowTok">if</span> (used.find(i) == used.end()) {
                <span class="cf" title="ControlFlowTok">if</span> (toupper(vals[i].first) == c || toupper(vals[i].second) == c) {
                    used.insert(i);
                    <span class="cf" title="ControlFlowTok">break</span>;
                }
            }
        }
        <span class="cf" title="ControlFlowTok">if</span> (x == used.size()) <span class="cf" title="ControlFlowTok">break</span>;
    }
    <span class="cf" title="ControlFlowTok">return</span> used.size() == w.size();
}


<span class="dt" title="DataTypeTok">int</span> main() {
    <span class="dt" title="DataTypeTok">list_t</span> vals{ {<span class="st" title="StringTok">&#39;B&#39;</span>,<span class="st" title="StringTok">&#39;O&#39;</span>}, {<span class="st" title="StringTok">&#39;X&#39;</span>,<span class="st" title="StringTok">&#39;K&#39;</span>}, {<span class="st" title="StringTok">&#39;D&#39;</span>,<span class="st" title="StringTok">&#39;Q&#39;</span>}, {<span class="st" title="StringTok">&#39;C&#39;</span>,<span class="st" title="StringTok">&#39;P&#39;</span>}, {<span class="st" title="StringTok">&#39;N&#39;</span>,<span class="st" title="StringTok">&#39;A&#39;</span>}, {<span class="st" title="StringTok">&#39;G&#39;</span>,<span class="st" title="StringTok">&#39;T&#39;</span>}, {<span class="st" title="StringTok">&#39;R&#39;</span>,<span class="st" title="StringTok">&#39;E&#39;</span>}, {<span class="st" title="StringTok">&#39;T&#39;</span>,<span class="st" title="StringTok">&#39;G&#39;</span>}, {<span class="st" title="StringTok">&#39;Q&#39;</span>,<span class="st" title="StringTok">&#39;D&#39;</span>}, {<span class="st" title="StringTok">&#39;F&#39;</span>,<span class="st" title="StringTok">&#39;S&#39;</span>}, {<span class="st" title="StringTok">&#39;J&#39;</span>,<span class="st" title="StringTok">&#39;W&#39;</span>}, {<span class="st" title="StringTok">&#39;H&#39;</span>,<span class="st" title="StringTok">&#39;U&#39;</span>}, {<span class="st" title="StringTok">&#39;V&#39;</span>,<span class="st" title="StringTok">&#39;I&#39;</span>}, {<span class="st" title="StringTok">&#39;A&#39;</span>,<span class="st" title="StringTok">&#39;N&#39;</span>}, {<span class="st" title="StringTok">&#39;O&#39;</span>,<span class="st" title="StringTok">&#39;B&#39;</span>}, {<span class="st" title="StringTok">&#39;E&#39;</span>,<span class="st" title="StringTok">&#39;R&#39;</span>}, {<span class="st" title="StringTok">&#39;F&#39;</span>,<span class="st" title="StringTok">&#39;S&#39;</span>}, {<span class="st" title="StringTok">&#39;L&#39;</span>,<span class="st" title="StringTok">&#39;Y&#39;</span>}, {<span class="st" title="StringTok">&#39;P&#39;</span>,<span class="st" title="StringTok">&#39;C&#39;</span>}, {<span class="st" title="StringTok">&#39;Z&#39;</span>,<span class="st" title="StringTok">&#39;M&#39;</span>} };
    <span class="bu" title="BuiltInTok">std::</span>vector&lt;<span class="bu" title="BuiltInTok">std::</span>string&gt; words{<span class="st" title="StringTok">&quot;A&quot;</span>,<span class="st" title="StringTok">&quot;BARK&quot;</span>,<span class="st" title="StringTok">&quot;BOOK&quot;</span>,<span class="st" title="StringTok">&quot;TREAT&quot;</span>,<span class="st" title="StringTok">&quot;COMMON&quot;</span>,<span class="st" title="StringTok">&quot;SQUAD&quot;</span>,<span class="st" title="StringTok">&quot;CONFUSE&quot;</span>};
    <span class="cf" title="ControlFlowTok">for</span> (<span class="at" title="AttributeTok">const</span> <span class="bu" title="BuiltInTok">std::</span>string&amp; w : words) {
        <span class="bu" title="BuiltInTok">std::</span>cout &lt;&lt; w &lt;&lt; <span class="st" title="StringTok">&quot;: &quot;</span> &lt;&lt; <span class="bu" title="BuiltInTok">std::</span>boolalpha &lt;&lt; can_make_word(w,vals) &lt;&lt; <span class="st" title="StringTok">&quot;.</span><span class="sc" title="SpecialCharTok">\n</span><span class="st" title="StringTok">&quot;</span>;
    }

}</code></pre></div>
